velox/functions/prestosql/GeometryUtils.h

```diff
@@ -0,0 +1,685 @@
+
+    void writeLong(int64_t value) {
+      write<int64_t>(value);
+    }
+
+    void writeFloat(float value) {
+      write<float>(value);
+    }
+
+    void writeDouble(double value) {
+      write<double>(value);
+    }
+
+    void writeBytes(const char* data, size_t size) {
+      stringWriter_.append(std::string_view(data, size));
+    }
+
+   private:
+    template <typename T>
+    void write(const T& value) {
+      stringWriter_.append(
+          std::string_view(reinterpret_cast<const char*>(&value), sizeof(T)));
+    }
+
+    StringWriter& stringWriter_;
+  };
+
+  template <typename T>
+  static void serialize(
+      const geos::geom::Geometry* geometry,
+      SliceOutput<T>& sliceOutput) {
+    auto geometryType = geometry->getGeometryTypeId();
+    switch (geometryType) {
+      case geos::geom::GEOS_POINT:
+        writePoint(geometry, sliceOutput);
+        break;
+      case geos::geom::GEOS_MULTIPOINT:
+        writeMultiPoint(geometry, sliceOutput);
+        break;
+      case geos::geom::GEOS_LINESTRING:
+      case geos::geom::GEOS_LINEARRING:
+        writePolyline(geometry, sliceOutput, false);
+        break;
+      case geos::geom::GEOS_MULTILINESTRING:
+        writePolyline(geometry, sliceOutput, true);
+        break;
+      case geos::geom::GEOS_POLYGON:
+        writePolygon(geometry, sliceOutput, false);
+        break;
+      case geos::geom::GEOS_MULTIPOLYGON:
+        writePolygon(geometry, sliceOutput, true);
+        break;
+      case geos::geom::GEOS_GEOMETRYCOLLECTION:
+        writeGeometryCollection(geometry, sliceOutput);
+        break;
+      default:
+        VELOX_USER_FAIL(
+            "Unrecognized geometry type: {}", static_cast<int>(geometryType));
+        break;
+    }
+  }
+
+  static std::unique_ptr<geos::geom::Geometry> deserialize(
+      SliceInput& sliceInput) {
+    auto geometryType =
+        static_cast<GeometrySerializationType>(sliceInput.readByte());
+    switch (geometryType) {
+      case GeometrySerializationType::POINT:
+        return readPoint(sliceInput);
+      case GeometrySerializationType::MULTI_POINT:
+        return readMultiPoint(sliceInput);
+      case GeometrySerializationType::LINE_STRING:
+        return readPolyline(sliceInput, false);
+      case GeometrySerializationType::MULTI_LINE_STRING:
+        return readPolyline(sliceInput, true);
+      case GeometrySerializationType::POLYGON:
+        return readPolygon(sliceInput, false);
+      case GeometrySerializationType::MULTI_POLYGON:
+        return readPolygon(sliceInput, true);
+      case GeometrySerializationType::ENVELOPE:
+        return readEnvelope(sliceInput);
+      case GeometrySerializationType::GEOMETRY_COLLECTION:
+        return readGeometryCollection(sliceInput);
+      default:
+        VELOX_USER_FAIL(
+            "Unrecognized geometry type: {}",
+            static_cast<uint8_t>(geometryType));
+        break;
+    }
+    return nullptr;
+  }
+
+  static bool isEsriNaN(double d) {
```

Comment:
nit: Is there a special NaN encoding for Geometric types ??

Reply:
Yes, ESRI has a magic float value that it considers a NaN.  I don't know why, but I could guess to have decent equality/ordering semantics.  I don't want to support that long term (as we move of ESRI), but for the first stage we should stay byte-wise compatible with the format, because the optimizer might do constant evaluation with the Java code (which uses the old format).